---
title: "Figures"
author: "Kenneth Daily"
date: "April 6, 2016"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
# Load required libraries
library(synapseClient)
library(ggplot2)
library(RColorBrewer)
library(gplots)
library(edgeR)
library(reshape2)
library(data.table)
library(plyr)
library(dplyr)
library(knitr)
library(stringr)
library(pcbcStats)

## Needed for vcd::assocstats to compute Cramer's V
library(vcd)
## Needs the dev branch
library(githubr)

library(ComplexHeatmap)
library(circlize)

# Simple covariate correlation function
# returns Cramer's V and p-value of a chi square test
# Drops any rows with NA values prior to computing
cov.corr.fxn <- function(Var1, Var2) {
  
  tmp <- na.omit(covariates[, c(Var1, Var2)])
  
  fac1 <- factor(tmp[, Var1])
  fac2 <- factor(tmp[, Var2])
  
  stats = assocstats(xtabs(~fac1 + fac2))
  
  return(data.frame(cv=stats$cramer,
                    p.value=stats$chisq_tests['Pearson','P(> X^2)']))
}

# Computes pairwise column correlations using Cramer's V
factorCramersV <- function(x, p) {
  cov_comparisons <- expand.grid(colnames(x), colnames(x))
  res <- ldply(apply(cov_comparisons, 1, 
                     function(x) {
                       cov.corr.fxn(Var1=x['Var1'],
                                     Var2=x['Var2'])
                     }))
  
  cov_comparisons <- cbind(cov_comparisons, res)
  cov_comparisons
}

synapseLogin()

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = FALSE)

set.seed(1234567890)
```

```{r include=TRUE, echo=FALSE}
# Synapse IDs of data matrix and metadata tables
mrnaDataId <- 'syn5008587'
mrnaMetadataId <- 'syn3156503'

mirnaDataId <- 'syn5014443'
mirnaMetadataId <- 'syn3219876'
mirnaReadsId = "syn5014445"

methylDataId = "syn2233188"
methylMetadataId = "syn3156828"
```

```{r mrnadata}
# Get count matrix
count_obj <- synGet(mrnaDataId)
countData = fread(getFileLocation(count_obj), 
                  data.table=FALSE, showProgress=FALSE)
row.names(countData) = countData[,1]
```

```{r mrnametadata}
# Get metadata
metadata_obj <- synTableQuery(paste('SELECT * FROM', mrnaMetadataId))
metadata <- metadata_obj@values

metadata[metadata == 'N/A'] <- NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
metadata <- metadata %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName,
                     -public, -pass_qc, -exclude) # fix them but don't touch some columns

# Set rownames
rownames(metadata) = metadata$UID
```

```{r mrnafiltering}
metadata_keep <- metadata %>% 
  filter(UID %in% colnames(countData))

metadata_filtered <- metadata_keep %>%
  filter(public, pass_qc, !exclude,
         Diffname_short != "",
         Cell_Type == "PSC",
         C4_Karyotype_Result != "abnormal")

metadata <- metadata_filtered
rownames(metadata) <- metadata$UID
countData <- countData[, c('hg19.kgXref.geneSymbol', metadata$UID)]

tmp <- tapply(metadata$UID,
              factor(metadata$Diffname_short),
              function(cols, countData) {
                mat <- DGEList(counts=countData[,cols])
                cpms <- cpm(mat)
                fracSamplesWithMinCPM <- rowMeans(cpms >= 1)
                isNonLowExpr <- fracSamplesWithMinCPM >= 0.5
                cpms[isNonLowExpr, ]
              },
              countData)


ugenes <- as.character(unlist(lapply(tmp, rownames)))

countData <- countData %>% filter(hg19.kgXref.geneSymbol %in% ugenes)

metadata.mrna <- metadata

```

```{r mrnapca, fig.height=8, fig.width=8}
voom_gene_expression <- voom(countData[, metadata$UID], design=NULL)

pc <- prcomp(voom_gene_expression$E, center=TRUE, scale=FALSE)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(pc$rotation), 
                       PC1=pc$rotation[,1], 
                       PC2=pc$rotation[,2])

var.pc1 <- summary(pc)$importance['Proportion of Variance', "PC1"] * 100
var.pc2 <- summary(pc)$importance['Proportion of Variance', "PC2"] * 100

plotdata <- merge(plotdata, metadata, by="UID")
plotdata <- dplyr::rename(plotdata, DifferentiationStage=Diffname_short)
plotdata$DifferentiationStage <- factor(plotdata$DifferentiationStage,
                                        levels=c("SC", "DE", "MESO5", "ECTO",
                                                 "MESO15", "MESO30", "EB"),
                                        ordered=TRUE)

pcamRNA <- ggplot(plotdata, aes(x=PC1, y=PC2))
pcamRNA <- pcamRNA + geom_point(aes(color=DifferentiationStage), size=3)
pcamRNA <- pcamRNA +  scale_color_manual(values=c("Black", "Red", "Blue",
                                                  "Green", "Purple",
                                                  "Orange", "Brown"),
                             guide=guide_legend(nrow=1, title=NULL))

pcamRNA <- pcamRNA + theme_bw() + theme(legend.position="top",
                                        text=element_text(size=16),
                                        title=element_text(size=16)) 
pcamRNA <- pcamRNA + labs(title = "mRNA",
                          x=sprintf("PC1 (%0.0f%%)", var.pc1),
                          y=sprintf("PC2 (%0.0f%%)", var.pc2))

```

```{r mirnadata}
mat_obj =  synGet(id=mirnaDataId)
# mat = fread(mat_obj@filePath, data.table=FALSE, showProgress=FALSE)
mat = read.table(mat_obj@filePath, check.names = F, header=T, sep="\t")

mat$featureName <- rownames(mat)

reads_obj =  synGet(id=mirnaReadsId)
reads = read.table(reads_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)
```

```{r mirnametadata}
metadata_obj = synTableQuery(paste('SELECT * FROM', mirnaMetadataId))
metadata = metadata_obj@values
metadata[metadata == 'N/A'] <- NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
metadata <- metadata %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName,
                     -public, -pass_qc, -exclude) # fix them but don't touch some columns

# Set rownames
rownames(metadata) = metadata$UID
```

```{r mirnafiltering}
#### Pre processing mRNA expression counts and metadata ####
metadata_keep <- metadata %>% 
  filter(UID %in% colnames(mat))

metadata_filtered <- metadata_keep %>%
  filter(public, # pass_qc, !exclude,
         Diffname_short != "",
         Cell_Type == "PSC",
         C4_Karyotype_Result != "abnormal")

metadata <- metadata_filtered
# rownames(metadata) <- metadata_filtered$UID
mat <- mat[, c('featureName', metadata$UID)]

reads = reads[metadata$UID,]

keep = reads %>% 
  mutate(samples=rownames(reads)) %>% 
  filter(mirna > 500000 & norm > 0.2)

metadata$size = reads[metadata$UID,"norm"]
metadata$size_cat = reads[metadata$UID,"breaks"]
metadata$mirna = log2(reads[metadata$UID, "mirna"])

mat <- mat[, c('featureName', keep$samples)]
metadata <- filter(metadata, UID %in% keep$samples)
rownames(metadata) <- metadata$UID

tmp <- tapply(metadata$UID,
              factor(metadata$Diffname_short),
              function(cols, countData) {
                mat <- DGEList(counts=countData[,cols])
                cpms <- cpm(mat)
                cpms[is.nan(cpms)] = 0
                fracSamplesWithMinCPM <- rowMeans(cpms >= 10)
                isNonLowExpr <- fracSamplesWithMinCPM >= 0.9
                cpms[isNonLowExpr, ]
              },
              mat)


ugenes <- unique(as.character(unlist(lapply(tmp, 
                                            function(x) rownames(x)))))

mat <- mat %>% filter(featureName %in% ugenes)

metadata.mirna <- metadata
```


```{r mirnapca, fig.height=8, fig.width=8}
voom_gene_expression <- voom(mat[, metadata$UID], design=NULL)

pc <- prcomp(voom_gene_expression$E, center=TRUE, scale=FALSE)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(pc$rotation), 
                       PC1=pc$rotation[,1], 
                       PC2=pc$rotation[,2])

var.pc1 <- summary(pc)$importance['Proportion of Variance', "PC1"] * 100
var.pc2 <- summary(pc)$importance['Proportion of Variance', "PC2"] * 100

plotdata <- merge(plotdata, metadata, by="UID")
plotdata <- dplyr::rename(plotdata, DifferentiationStage=Diffname_short)
plotdata$DifferentiationStage <- factor(plotdata$DifferentiationStage,
                                        levels=c("SC", "DE", "MESO5", "ECTO",
                                                 "MESO15", "MESO30", "EB"),
                                        ordered=TRUE)

pcamiRNA <- ggplot(plotdata, aes(x=PC1, y=PC2))
pcamiRNA <- pcamiRNA + geom_point(aes(color=DifferentiationStage), size=3)
pcamiRNA <- pcamiRNA +  scale_color_manual(values=c("Black", "Red", "Blue",
                                                    "Green", "Purple",
                                                    "Orange", "Brown"),
                                           guide=guide_legend(nrow=1, title=NULL))

pcamiRNA <- pcamiRNA + theme_bw() + theme(legend.position="top",
                                          text=element_text(size=16),
                                          title=element_text(size=16)) 
pcamiRNA <- pcamiRNA + labs(title = "microRNA",
                            x=sprintf("PC1 (%0.0f%%)", var.pc1),
                            y=sprintf("PC2 (%0.0f%%)", var.pc2))

```

```{r methyldata}
mat_obj =  synGet(id=methylDataId)
mat = fread(mat_obj@filePath, data.table=FALSE, showProgress=FALSE)
rownames(mat) <- mat$ProbeID
```

```{r methylmetadata}
metadata_obj = synTableQuery(paste('SELECT * FROM', methylMetadataId))
metadata = metadata_obj@values
metadata[metadata == 'N/A'] <- NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
metadata <- metadata %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName,
                     -public, -pass_qc, -exclude) # fix them but don't touch some columns

# Set rownames
rownames(metadata) = metadata$UID
```

```{r methylfiltering}
#### Pre processing mRNA expression counts and metadata ####
metadata_keep <- metadata %>% 
  filter(UID %in% colnames(mat))

metadata_filtered <- metadata_keep %>%
  filter(public,# pass_qc, !exclude,
         biologicalSampleName != "H9P50",
         Diffname_short != "",
         Cell_Type == "PSC",
         C4_Karyotype_Result != "abnormal")

metadata <- metadata_filtered
# rownames(metadata) <- metadata_filtered$UID
mat <- mat[, c('ProbeID', metadata$UID)]

metadata.methyl <- metadata

exp.variance <- apply(as.matrix(mat[, metadata$UID]), 1, var)
methyl.var.cutoff <- quantile(exp.variance)['25%'] 
ind.remove1 <- exp.variance < methyl.var.cutoff
#ind.remove1 <- exp.variance < sort(exp.variance)[round(0.25*length(exp.variance))]

nsamples <- ncol(mat[, metadata$UID])
ind.remove2 <- ((rowSums(mat[, metadata$UID] <= 0.25) / nsamples) == 1) | 
  ((rowSums(mat[, metadata$UID] >= 0.75) / nsamples) == 1)

mat <- mat[!ind.remove1 & !ind.remove2,]
```

```{r methylpca, fig.height=8, fig.width=8}
pc <- prcomp(mat[, metadata$UID], center=TRUE, scale=FALSE)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(pc$rotation), 
                       PC1=pc$rotation[,1], 
                       PC2=pc$rotation[,2])

var.pc1 <- summary(pc)$importance['Proportion of Variance', "PC1"] * 100
var.pc2 <- summary(pc)$importance['Proportion of Variance', "PC2"] * 100

plotdata <- merge(plotdata, metadata, by="UID")
plotdata <- dplyr::rename(plotdata, DifferentiationStage=Diffname_short)
plotdata$DifferentiationStage <- factor(plotdata$DifferentiationStage,
                                        levels=c("SC", "DE", "MESO5", "ECTO",
                                                 "MESO15", "MESO30", "EB"),
                                        ordered=TRUE)

pcaMethyl <- ggplot(plotdata, aes(x=PC1, y=PC2))
pcaMethyl <- pcaMethyl + geom_point(aes(color=DifferentiationStage), size=3)
pcaMethyl <- pcaMethyl +  scale_color_manual(values=c("Black", "Red", "Blue", 
                                                      "Green", "Brown"),
                                             guide=guide_legend(nrow=1, title=NULL))

pcaMethyl <- pcaMethyl + theme_bw() + theme(legend.position="top",
                                            text=element_text(size=16),
                                            title=element_text(size=16))
pcaMethyl <- pcaMethyl + labs(title = "Methylation", 
                              x=sprintf("PC1 (%0.0f%%)", var.pc1),
                              y=sprintf("PC2 (%0.0f%%)", var.pc2))
```

### Covariate clustering
```{r covclustering}
factor_covariates <- c(#'Diffname_short', 
                       'Cell_Line_Type',
                       'Cell_Line_of_Origin', 'Tissue_of_Origin',
                       'Reprogramming_Gene_Combination', 'Culture_Conditions',
                       'Donor_Life_Stage', 'Gender', 'Originating_Lab', 'Donor_ID',
                       'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')

metadata.merge <- rbind(metadata.mrna[, c("C4_Cell_Line_ID", factor_covariates)],
                        metadata.mirna[, c("C4_Cell_Line_ID", factor_covariates)],
                        metadata.methyl[, c("C4_Cell_Line_ID", factor_covariates)]) %>% unique

covariates = metadata.merge[, factor_covariates]

# Convert factor covariates to factors
covariates[,factor_covariates] = lapply(covariates[,factor_covariates], 
                                        factor)

cov_comparisons <- factorCramersV(covariates[, factor_covariates])

cramer_mat <- reshape2::dcast(cov_comparisons,# %>% filter(p.value < 0.1), 
                              Var1 ~ Var2, value.var="cv", 
                              fill=0)

rownames(cramer_mat) <- cramer_mat$Var1
cramer_mat$Var1 <- NULL
```

We examined the cell line covariates for potential correlations. All variables included are factors and Cramer's V was used as the statistic. The resulting pairwise values were clustered using hierarchical clustering using Ward's method with Euclidean distance. The variables included are `r cat(paste(gsub('_','\\\\_',factor_covariates), collapse=', '))`.

```{r covheatmap, fig.width=10, fig.height=10}

Heatmap(as.matrix(cramer_mat), name="Cramer's V",
        col=brewer.pal(9, "Reds"),
        clustering_method_rows="ward.D",
        clustering_method_columns="ward.D",
        column_names_max_height = unit(6, "cm"),
        row_names_max_width = unit(6, "cm"),
        show_row_dend=FALSE)

```

### Filtering cell lines
We remove sample lines with abnormal karyotypes (syn2679104). We also removed samples that failed quality control at the time of assay (low RIN, etc.) and those that were flagged as problematic after alignment (mRNA and miRNA) or microarray analysis (methylation).

### Filtering assay data

#### mRNA
We use the effective count matrix generated from eXpress and collapsed it from transcripts to genes by summing the counts (`r mrnaDataId`). We remove genes that have less than 1 count per million (CPM) in more than half of the samples within each individual differentiation stage.

#### miRNA
We use the effective count matrix generated from `seqbuster` (`r mirnaDataId`). We remove samples that have less than 500000 aligned reads or have less than 20% of the total reads aligned. We remove miRNAs that have less than 10 CPM in at least 90% of each of the individual differentiation stages.

#### Methylation
We use the probe-level beta matrix from `minfi` (`r methylDataId`). We remove probes with beta value variance lower than the 25th percentile (`r methyl.var.cutoff`) or with beta values all below 0.25 or all above 0.75.

### Principal components analysis
We use PCA to visualize the samples in each assay using centered, non-scaled data. We plot the first two principal components and color the points by the differentiation stage. As expected, the samples generally cluster by differentiation stage, indicating this variable is a strong driver of the variability in the datasets.

```{r fig.width=20, fig.height=10}
multiplot(pcamRNA, pcamiRNA, pcaMethyl, cols=3)

png("pca_plots.png", width = 1500, height = 500)
multiplot(pcamRNA, pcamiRNA, pcaMethyl, cols=3)
foo <- dev.off()
```

